cli/src/commands/extensions/extension.rs

```diff
@@ -0,0 +1,167 @@
+use std::convert::TryFrom;
+use std::fs::{self, Permissions};
+#[cfg(unix)]
+use std::os::unix::prelude::PermissionsExt;
+use std::path::PathBuf;
+
+use anyhow::{anyhow, Result};
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::Deserialize;
+use walkdir::WalkDir;
+
+const MANIFEST_NAME: &str = "PhylumExt.toml";
+
+lazy_static! {
+    static ref EXTENSION_NAME_RE: Regex = Regex::new(r#"^[a-z0-9_-]+$"#).unwrap();
+}
+
+#[derive(Debug)]
+pub struct Extension {
+    path: PathBuf,
+    manifest: ExtensionManifest,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct ExtensionManifest {
+    name: String,
+    description: Option<String>,
+    entry_point: String,
+}
+
+impl Extension {
+    pub fn name(&self) -> &str {
+        &self.manifest.name
+    }
+
+    pub fn description(&self) -> Option<&str> {
+        self.manifest.description.as_deref()
+    }
+
+    /// Install the extension in the default path.
+    pub fn install(&self) -> Result<()> {
+        println!("Installing extension {}...", self.name());
+
+        let target_prefix = extension_path(self.name())?;
+
+        // TODO we may want to implement `upgrade` in the future, which would
+        // allow writing to the path of an already installed extension.
+        if target_prefix.exists() {
+            return Err(anyhow!("extension already exists, skipping"));
+        }
+
+        if target_prefix == self.path {
+            return Err(anyhow!(
+                "extension path and installation path are identical, skipping"
+            ));
+        }
+
+        for entry in WalkDir::new(&self.path) {
+            let source_path = entry?.into_path();
+            let dest_path = target_prefix.join(source_path.strip_prefix(&self.path)?);
+
+            if source_path.is_dir() {
+                fs::create_dir_all(&dest_path)?;
+                #[cfg(unix)]
+                fs::set_permissions(&dest_path, Permissions::from_mode(0o700))?;
```

Comment:
This would be more pertinent to #353. I think checking specifically for XDG would add unnecessary complexity, but we can consider it if there are strong reasons why we may want to explicitly not set permissions.